File: src/lib/layerb.ts
import type { LayerBResult, RelianceLabel, RelianceMode, SegmentCoding, Trace, TraceTurn } from '../types';
import type { LLMClient } from './llm/client';
import { SEGMENT_SCHEMA, buildSegmentPrompt, buildSegmentSystem } from './grading/prompts';

/** Split the dialogue into segments: each student turn plus the assistant reply that
 *  follows it (and the assistant turn it responds to, for context). */
export function segmentTrace(trace: Trace): TraceTurn[][] {
  const segments: TraceTurn[][] = [];
  const turns = trace.turns;
  for (let i = 0; i < turns.length; i++) {
    if (turns[i].speaker !== 'student') continue;
    const seg: TraceTurn[] = [];
    if (i > 0 && turns[i - 1].speaker === 'assistant') seg.push(turns[i - 1]);
    seg.push(turns[i]);
    if (i + 1 < turns.length && turns[i + 1].speaker === 'assistant') seg.push(turns[i + 1]);
    segments.push(seg);
  }
  return segments;
}

const MODES: RelianceMode[] = ['passive', 'active', 'constructive'];

export function summarizeSegments(segments: SegmentCoding[]): LayerBResult {
  const grid = {} as LayerBResult['grid'];
  for (const h of MODES) {
    grid[h] = { passive: 0, active: 0, constructive: 0 };
  }
  for (const s of segments) grid[s.helpSeeking][s.responseUse]++;

  const count = (dim: 'helpSeeking' | 'responseUse', m: RelianceMode) => segments.filter((s) => s[dim] === m).length;
  const dominant = (dim: 'helpSeeking' | 'responseUse'): RelianceMode =>
    MODES.reduce((best, m) => (count(dim, m) > count(dim, best) ? m : best), 'passive' as RelianceMode);

  const dominantHelpSeeking = dominant('helpSeeking');
  const dominantResponseUse = dominant('responseUse');
  const verificationRate = segments.length ? segments.filter((s) => s.verification).length / segments.length : 0;

  // Interpretive label heuristic per Hou et al. (2025): shown as a hypothesis, not a verdict.
  let interpretiveLabel: RelianceLabel;
  if (dominantResponseUse === 'passive' && verificationRate < 0.2) interpretiveLabel = 'thoughtless';
  else if (dominantResponseUse === 'constructive' && dominantHelpSeeking === 'constructive') interpretiveLabel = 'collaborative';
  else if (verificationRate >= 0.5) interpretiveLabel = 'reflective';
  else interpretiveLabel = 'cautious';

  return { segments, grid, dominantHelpSeeking, dominantResponseUse, interpretiveLabel, verificationRate };
}

export async function codeLayerB(
  llm: LLMClient,
  trace: Trace,
  onProgress?: (done: number, total: number) => void,
): Promise<LayerBResult> {
  const rawSegments = segmentTrace(trace);
  const codings: SegmentCoding[] = [];
  let done = 0;
  for (const seg of rawSegments) {
    const text = seg.map((t) => `[turn ${t.turnId} | ${t.speaker.toUpperCase()}]\n${t.text}`).join('\n\n');
    const raw = (await llm.completeJSON({
      system: buildSegmentSystem(),
      prompt: buildSegmentPrompt(text),
      schema: SEGMENT_SCHEMA,
    })) as Partial<SegmentCoding>;
    codings.push({
      segmentTurns: seg.map((t) => t.turnId),
      helpSeeking: (raw.helpSeeking as RelianceMode) ?? 'active',
      responseUse: (raw.responseUse as RelianceMode) ?? 'active',
      verification: Boolean(raw.verification),
      evidence: raw.evidence ?? '',
    });
    done++;
    onProgress?.(done, rawSegments.length);
  }
  return summarizeSegments(codings);
}

File: backend/app/services/grading/layerb.py
"""Layer B: RelianceScope 3×3 AI-reliance coding — port of TGFWA src/lib/layerb.ts.

Layer B describes HOW the student worked with the AI (help-seeking × response-use
per segment, verification behavior, interpretive label per Hou et al. 2025).
It is never blended with Layer A writing-mastery scores.
"""

from .prompts import build_segment_prompt, build_segment_system

MODES = ("passive", "active", "constructive")


def segment_trace(trace: dict) -> list:
    """Split the dialogue into segments: each student turn plus the assistant reply
    that follows it (and the assistant turn it responds to, for context)."""
    segments = []
    turns = trace.get("turns", [])
    for i, turn in enumerate(turns):
        if turn.get("speaker") != "student":
            continue
        seg = []
        if i > 0 and turns[i - 1].get("speaker") == "assistant":
            seg.append(turns[i - 1])
        seg.append(turn)
        if i + 1 < len(turns) and turns[i + 1].get("speaker") == "assistant":
            seg.append(turns[i + 1])
        segments.append(seg)
    return segments


def summarize_segments(segments: list) -> dict:
    grid = {h: {m: 0 for m in MODES} for h in MODES}
    for s in segments:
        grid[s["helpSeeking"]][s["responseUse"]] += 1

    def count(dim, mode):
        return sum(1 for s in segments if s[dim] == mode)

    def dominant(dim):
        best = "passive"
        for m in MODES:
            if count(dim, m) > count(dim, best):
                best = m
        return best

    dominant_help_seeking = dominant("helpSeeking")
    dominant_response_use = dominant("responseUse")
    verification_rate = (sum(1 for s in segments if s["verification"]) / len(segments)
                         if segments else 0)

    # Interpretive label heuristic per Hou et al. (2025): a hypothesis, not a verdict.
    if dominant_response_use == "passive" and verification_rate < 0.2:
        label = "thoughtless"
    elif dominant_response_use == "constructive" and dominant_help_seeking == "constructive":
        label = "collaborative"
    elif verification_rate >= 0.5:
        label = "reflective"
    else:
        label = "cautious"

    return {
        "segments": segments,
        "grid": grid,
        "dominantHelpSeeking": dominant_help_seeking,
        "dominantResponseUse": dominant_response_use,
        "interpretiveLabel": label,
        "verificationRate": verification_rate,
    }


def code_layer_b(llm_json, trace: dict, on_progress=None) -> dict:
    raw_segments = segment_trace(trace)
    codings = []
    for done, seg in enumerate(raw_segments, start=1):
        text = "\n\n".join(f"[turn {t['turnId']} | {t['speaker'].upper()}]\n{t['text']}"
                           for t in seg)
        raw = llm_json(build_segment_system(), build_segment_prompt(text))
        raw = raw if isinstance(raw, dict) else {}
        codings.append({
            "segmentTurns": [t["turnId"] for t in seg],
            "helpSeeking": raw.get("helpSeeking") if raw.get("helpSeeking") in MODES else "active",
            "responseUse": raw.get("responseUse") if raw.get("responseUse") in MODES else "active",
            "verification": bool(raw.get("verification")),
            "evidence": raw.get("evidence", ""),
        })
        if on_progress:
            on_progress(done, len(raw_segments))
    return summarize_segments(codings)
